[WebAuthn] Should reject rp with empty id

pascoej opened the pull request:
#### 3b920f82563c500e5c4991cc04bf0b99b784b9c2
<pre>
[WebAuthn] Should reject rp with empty id
<a href="https://bugs.webkit.org/show_bug.cgi?id=242091">https://bugs.webkit.org/show_bug.cgi?id=242091</a>
rdar://90281653

Reviewed by Brent Fulgham.

Not specifying an rp.id should default to the caller’s origin&apos;s effective domain, but
empty / null values should be rejected per spec. <a href="https://www.w3.org/TR/webauthn-2/#sctn-createCredential">https://www.w3.org/TR/webauthn-2/#sctn-createCredential</a>

Updated tests.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https.html:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create const):
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::convertRpEntityToCBOR):
* Source/WebCore/Modules/webauthn/fido/U2fCommandConstructor.cpp:
(fido::convertToU2fRegisterCommand):
(fido::convertToU2fCheckOnlySignCommand):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(publicKeyCredentialRpEntity):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::getRpId):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::configureRegistrationRequestContext):
* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::continueRegisterCommandAfterResponseReceived):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/252142@main">https://commits.webkit.org/252142@main</a>
</pre>

brentfulgham commented on LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https-expected.txt:
```diff
@@ -7,6 +7,8 @@ CONSOLE MESSAGE: User gesture is not detected. To use the WebAuthn API, call 'na
 PASS PublicKeyCredential's [[create]] with timeout
 PASS PublicKeyCredential's [[create]] with a mismatched RP ID
+PASS PublicKeyCredential's [[create]] with a empty string RP ID
+PASS PublicKeyCredential's [[create]] with a null RP ID
```
Will we eventually get 'LayoutTests/imported/w3c/web-platform-tests' to include the web authentication tests? Or are they all being imported to http? I'm not sure what our policy is on where WPT tests live.

pascoej replied:
I've been investigating using 'LayoutTests/imported/w3c/web-platform-tests' but haven't yet been successful in getting the tests to run within run-webkit-tests after import.

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
```diff
@@ -157,7 +157,7 @@ void U2fAuthenticator::continueRegisterCommandAfterResponseReceived(ApduResponse
         auto& options = std::get<PublicKeyCredentialCreationOptions>(requestData().options);
         auto appId = processGoogleLegacyAppIdSupportExtension(options.extensions);
-        auto response = readU2fRegisterResponse(!appId ? options.rp.id : appId, apduResponse.data(), AuthenticatorAttachment::CrossPlatform, { driver().transport() }, options.attestation);
+        auto response = readU2fRegisterResponse(!appId ? *options.rp.id : appId, apduResponse.data(), AuthenticatorAttachment::CrossPlatform, { driver().transport() }, options.attestation);
```
Since `options.rp.id` may now be `std::nullopt`, do we need to do any validation here? Maybe at least ASSERT so we catch cases where a null opt got through?

pascoej replied:
It should get replaced in AuthenticatorCoordinator when `std::nullopt`. I will add ASSERTs.

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -228,7 +228,7 @@ static inline ASPublicKeyCredentialResidentKeyPreference toASCResidentKeyPrefere
 
     auto requestContext = adoptNS([allocASCCredentialRequestContextInstance() initWithRequestTypes:requestTypes]);
-    [requestContext setRelyingPartyIdentifier:options.rp.id];
+    [requestContext setRelyingPartyIdentifier:*options.rp.id];
```
Since `options.rp.id` may now be `std::nullopt`, do we need to do any validation here? Maybe at least ASSERT so we catch cases where a null opt got through?

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
```diff
@@ -248,7 +248,7 @@ static inline uint8_t authDataFlags(ClientDataType type, LocalConnection::UserVe
 
     // Step 3.
-    auto existingCredentials = getExistingCredentials(creationOptions.rp.id);
+    auto existingCredentials = getExistingCredentials(*creationOptions.rp.id);
```
Since `options.rp.id` may now be `std::nullopt`, do we need to do any validation here? Maybe at least ASSERT so we catch cases where a null opt got through?

brentfulgham commented on Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h:
```diff
@@ -48,7 +48,7 @@ struct PublicKeyCredentialCreationOptions {
 
     struct RpEntity : public Entity {
-        mutable String id;
+        mutable std::optional<String> id;
```
I think program behavior is undefined if you dereference `id` when it is in a `std::null opt` state. I'm worried that the current change doesn't check for this possibility, which might lead to undefined program behavior.

pascoej committed:
[WebAuthn] Should reject rp with empty id
https://bugs.webkit.org/show_bug.cgi?id=242091
rdar://90281653

Reviewed by Brent Fulgham.

Not specifying an rp.id should default to the caller’s origin's effective domain, but
empty / null values should be rejected per spec. https://www.w3.org/TR/webauthn-2/#sctn-createCredential

Updated tests.

* LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-create-failure.https.html:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::create const):
* Source/WebCore/Modules/webauthn/PublicKeyCredentialCreationOptions.h:
* Source/WebCore/Modules/webauthn/fido/DeviceRequestConverter.cpp:
(fido::convertRpEntityToCBOR):
* Source/WebCore/Modules/webauthn/fido/U2fCommandConstructor.cpp:
(fido::convertToU2fRegisterCommand):
(fido::convertToU2fCheckOnlySignCommand):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebAuthenticationPanel.mm:
(publicKeyCredentialRpEntity):
* Source/WebKit/UIProcess/WebAuthentication/AuthenticatorManager.cpp:
(WebKit::WebCore::getRpId):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::makeCredential):
(WebKit::LocalAuthenticator::continueMakeCredentialAfterUserVerification):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::configureRegistrationRequestContext):
* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::continueRegisterCommandAfterResponseReceived):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKWebAuthenticationPanel.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/252142@main
